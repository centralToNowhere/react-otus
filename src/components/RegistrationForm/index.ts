export { RegistrationForm } from "./RegistrationForm";
export { RegistrationFormContainer } from "./RegistrationFormContainer";
export type { IRegistrationFormProps } from "./RegistrationForm";
