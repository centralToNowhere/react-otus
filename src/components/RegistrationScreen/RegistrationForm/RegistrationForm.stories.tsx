import React from "react";
import { Story, Meta } from "@storybook/react";
import {
  RegistrationForm,
  IRegistrationFormProps,
} from "@/components/RegistrationScreen/RegistrationForm";

export default {
  title: "Components/RegistrationScreen/RegistrationForm",
  component: RegistrationForm,
  args: {
    player: {
      registered: true,
      name: "Player1",
    },
    loginPending: false,
    loginError: "",
    onPlayerRegistration: () => {
      //empty
    },
  },
} as Meta;

const Template: Story<IRegistrationFormProps> = (args) => {
  return <RegistrationForm {...args} />;
};

export const PlayerRegistrationFormDefault = Template;

export const PlayerRegistrationFormPending = Template.bind({});
PlayerRegistrationFormPending.args = {
  loginPending: true,
};

export const PlayerRegistrationFormError = Template.bind({});
PlayerRegistrationFormError.args = {
  loginError: "Player is not registered!",
};
