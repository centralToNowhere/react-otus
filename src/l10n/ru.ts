export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
