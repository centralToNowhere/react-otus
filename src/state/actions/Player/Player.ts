export interface IPlayer {
  registered: boolean,
  name: string | null
}

export const isPlayer = (player: IPlayer): player is IPlayer => {
  return player && typeof player === "object"
    && typeof player.registered === "boolean"
    && (typeof player.name === "string" || player.name === null);
}