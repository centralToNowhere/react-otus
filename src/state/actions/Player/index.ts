export * from "./Player";
export * from "./SetPlayerAction";
