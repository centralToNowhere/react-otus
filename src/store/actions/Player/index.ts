export * from "./Player";
export * from "./SetPlayerAction";
export {IPlayer} from "@/player/Player";
