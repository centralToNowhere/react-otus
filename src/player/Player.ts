import { IFieldControlState } from "@/components/Fields";
import { IAuthState } from "@/auth";

export interface IPlayer {
  registered: boolean;
  name: string | null;
}

export const defaultPlayer: IPlayer = {
  registered: false,
  name: null,
};

export type PlayerStorageData = IFieldControlState &
  Omit<IAuthState, "loginPending">;
