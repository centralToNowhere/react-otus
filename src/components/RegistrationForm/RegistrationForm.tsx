import React, {
  FC,
  useCallback,
  useState,
  SyntheticEvent,
  KeyboardEvent,
} from "react";
import { l10n } from "@/l10n/ru";
import { FormField } from "@/components/Form/FormField";
import { FormElement } from "@/components/Form";
import { css } from "@emotion/react";
import { InputField, LabelField } from "@/components/Fields";
import { FormButton } from "@/components/Buttons";
import { BREAKPOINTS, COLORS } from "@/styles/ui-styled";
import styled from "@emotion/styled";
import { Spinner } from "react-bootstrap";
import { IPlayer } from "@/player/Player";
import { FieldError } from "@/components/Fields/FieldError";

export interface IRegistrationFormProps {
  player: IPlayer;
  loginPending: boolean;
  loginError: string;
  onPlayerRegistration: (playerName: string | null) => void;
}

export const RegistrationForm: FC<IRegistrationFormProps> = (props) => {
  const [playerName, setPlayerName] = useState(props.player.name);

  const onPlayerNameChange = useCallback(
    (e: SyntheticEvent<HTMLInputElement>) => {
      setPlayerName(e.currentTarget.value);
    },
    [setPlayerName]
  );

  const onKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      e.preventDefault();
      registerPlayer();
    }
  };

  const registerPlayer = () => {
    props.onPlayerRegistration(playerName);
  };

  return (
    <FormElement
      css={css`
        width: 100%;
        max-width: 600px;
        flex-direction: row;
        font-size: 1.2rem;
        margin: 0 0 20px 0;

        @media screen and (max-width: ${BREAKPOINTS.mobileEnd}) {
          flex-direction: column-reverse;
        }
      `}
    >
      <FormField
        css={css`
          display: flex;
          flex-grow: 10;
        `}
      >
        <div
          css={css`
            margin-bottom: 20px;
          `}
        >
          <LabelField htmlFor="register-player-input">
            {l10n.registerPlayerLabel}
          </LabelField>
          <InputField
            css={css`
              width: 100%;
            `}
            autoComplete={"off"}
            id={"register-player-input"}
            value={playerName || ""}
            onChange={onPlayerNameChange}
            name="playerName"
            autoFocus={true}
            onKeyDown={onKeyDown}
          />
          <FieldError show={Boolean(props.loginError)} msg={props.loginError} />
        </div>
        <FormButton
          type="button"
          css={css`
            margin-right: auto !important;
          `}
          name="startGameAsPlayer"
          onClick={registerPlayer}
        >
          {l10n.buttonStartGameAsPlayer}
        </FormButton>
      </FormField>
      <SpinnerContainer>
        <Spinner
          animation="border"
          role="status"
          css={css`
            color: ${COLORS.accent};
            visibility: ${props.loginPending ? "visible" : "hidden"};
            margin: 20px 20px 0 20px;

            @media screen and (max-width: ${BREAKPOINTS.mobileEnd}) {
              display: ${props.loginPending ? "inline-block" : "none"};
            }
          `}
        />
      </SpinnerContainer>
    </FormElement>
  );
};

const SpinnerContainer = styled.div`
  flex-grow: 1;
  text-align: right;

  @media screen and (max-width: ${BREAKPOINTS.mobileEnd}) {
    text-align: left;
  }
`;
