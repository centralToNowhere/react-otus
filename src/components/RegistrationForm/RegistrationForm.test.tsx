import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { RegistrationForm } from "@/components/RegistrationForm/RegistrationForm";
import { l10n } from "@/l10n/ru";
import userEvent from "@testing-library/user-event/dist";
import { defaultPlayer } from "@/player/Player";

describe("PlayerRegistrationForm tests", () => {
  it("should render player registration form", () => {
    const { asFragment } = render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={jest.fn()}
        loginPending={false}
        loginError=""
      />
    );

    const input = screen.getByLabelText(l10n.registerPlayerLabel);
    const button = screen.getByRole("button");

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onPlayerRegistration on button click", async () => {
    const onPlayerRegistration = jest.fn();

    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={onPlayerRegistration}
        loginPending={false}
        loginError=""
      />
    );

    const button = screen.getByRole("button");

    userEvent.click(button);

    await waitFor(() => {
      expect(onPlayerRegistration).toHaveBeenCalledTimes(1);
    });
  });

  it("should call onPlayerRegistration on button enter", async () => {
    const onPlayerRegistration = jest.fn();

    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={onPlayerRegistration}
        loginPending={false}
        loginError=""
      />
    );

    const button = screen.getByRole("button");

    userEvent.type(button, "{enter}", {
      skipClick: true,
    });

    await waitFor(() => {
      expect(onPlayerRegistration).toHaveBeenCalledTimes(1);
    });
  });

  it("should call onPlayerRegistration on enter keydown", async () => {
    const onPlayerRegistration = jest.fn();

    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={onPlayerRegistration}
        loginPending={false}
        loginError=""
      />
    );

    const input = screen.getByLabelText(l10n.registerPlayerLabel);

    userEvent.type(input, "{enter}");

    await waitFor(() => {
      expect(onPlayerRegistration).toHaveBeenCalledTimes(1);
    });
  });

  it("should render spinner on loginPending", () => {
    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={jest.fn()}
        loginPending={true}
        loginError=""
      />
    );

    const spinner = screen.queryByRole("status");

    expect(spinner).toBeVisible();
  });

  it("should not render spinner when loginPending = false", () => {
    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={jest.fn()}
        loginPending={false}
        loginError=""
      />
    );

    const spinner = screen.queryByRole("status");

    expect(spinner).toBeNull();
  });

  it("should render error message", () => {
    render(
      <RegistrationForm
        player={defaultPlayer}
        onPlayerRegistration={jest.fn()}
        loginPending={false}
        loginError="something went wrong"
      />
    );

    const error = screen.queryByRole("alert");

    expect(error).toHaveTextContent("something went wrong");
  });
});
