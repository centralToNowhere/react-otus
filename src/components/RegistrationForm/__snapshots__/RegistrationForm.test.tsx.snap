// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerRegistrationForm tests should render player registration form 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 2px solid #6a6d75;
  border-radius: 10px;
  width: 100%;
  max-width: 600px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 1.2rem;
  margin: 0 0 20px 0;
}

.emotion-0 input,
.emotion-0 label,
.emotion-0 p {
  line-height: 1;
}

.emotion-0.game-settings-form .field-error-msg {
  background: #e8d5f0;
}

@media screen and (max-width: 480px) {
  .emotion-0 {
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
  }
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 calc(100% / 3 - 40px);
  -ms-flex: 0 0 calc(100% / 3 - 40px);
  flex: 0 0 calc(100% / 3 - 40px);
  margin: 20px;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 10;
  -webkit-flex-grow: 10;
  -ms-flex-positive: 10;
  flex-grow: 10;
}

.emotion-2 input,
.emotion-2 button {
  margin: 5px 0 0 0;
  min-width: 100px;
}

.emotion-4 {
  margin-bottom: 20px;
}

.emotion-5 {
  display: inline;
}

.emotion-7 {
  background: inherit;
  border: 0;
  border-bottom: 2px solid;
  border-color: #6a6d75;
  display: block;
  outline: 0;
  width: 100px;
  width: 100%;
}

.emotion-7:focus {
  border-bottom: 2px solid rgb(255, 33, 92);
}

.emotion-9 {
  background: inherit;
  margin: 20px auto;
  border-color: #6a6d75;
  line-height: 1.5;
  margin-right: auto!important;
}

.emotion-9:focus {
  outline: 2px solid rgb(255, 33, 92)!important;
  border: 2px solid transparent;
}

.emotion-11 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
}

@media screen and (max-width: 480px) {
  .emotion-11 {
    text-align: left;
  }
}

.emotion-13 {
  color: rgb(255, 33, 92);
  visibility: hidden;
  margin: 20px 20px 0 20px;
}

@media screen and (max-width: 480px) {
  .emotion-13 {
    display: none;
  }
}

<form
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <label
          class="emotion-5 emotion-6"
          for="register-player-input"
        >
          Введите имя игрока
        </label>
        <input
          autocomplete="off"
          class="emotion-7 emotion-8"
          id="register-player-input"
          name="playerName"
          value=""
        />
      </div>
      <button
        class="emotion-9 emotion-10"
        name="startGameAsPlayer"
        type="button"
      >
        Начать игру
      </button>
    </div>
    <div
      class="emotion-11 emotion-12"
    >
      <div
        class="emotion-13 spinner-border"
        role="status"
      />
    </div>
  </form>
</DocumentFragment>
`;
